### PR TITLE
Improve compile time

### LINKED ISSUE #1738: Unbearable linking speed when building on Mingw
When building the fc library on Mingw64/Msys2 platform, the linking process of api.exe takes unbearable time.

I'm trying to do some work for #964, starting with fc library first.  
It's almost done at [cwyyprog/bitshares-fc](/cwyyprog/bitshares-fc/tree/master),  the problem defeated me is that ld.exe works too slowly when linking api.exe and reaches the 1 hour limit of appveyor.

The appveror conf is :
``` yml
version: 1.0.{build}.mingw64
branches:
  only:
  - master
skip_tags: true
image: Visual Studio 2017
init:
- ps: $blockRdp = $false; iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/appveyor/ci/master/scripts/enable-rdp.ps1'))
clone_script:
- cmd: git clone --recursive https://github.com/cwyyprog/bitshares-fc.git
build_script:
- cmd: >-
    set MSYSTEM=MINGW64
    cd %APPVEYOR_BUILD_FOLDER%
    
    c:\msys64\usr\bin\bash -lc "pacman --noconfirm  -Syu"
    
    c:\msys64\usr\bin\bash -lc "pacman --noconfirm  -Syu"
    C:\msys64\usr\bin\bash -lc $APPVEYOR_BUILD_FOLDER/bitshares-fc/appveyor.bash
test: off
deploy: off
```
and the build script is:
``` bash
#!/usr/bin/bash
export MSYSTEM=MINGW64
pacman --noconfirm -S  mingw-w64-x86_64-boost; #mingw-w64-x86_64-openssl
mkdir /msys64
mount C:/msys64 /msys64;
cd $APPVEYOR_BUILD_FOLDER/bitshares-fc;
cmake -G "MSYS Makefiles" . -DOPENSSL_CONF_SOURCE="C:/msys64/mingw64/ssl/openssl.cnf" -DOPENSSL_ROOT_DIR=/mingw64 -DBOOST_ROOT=/mingw64;
#cmake --build . -j 2 --target all_tests;
cmake --build . -j 2 --target log_test;
cmake --build . -j 2 --target bloom_test;
cmake --build . -j 2 --target hmac_test;
cmake --build . -j 2 --target blind;
cmake --build . -j 2 --target task_cancel_test;
cmake --build . -j 2 --target api;
cmake --build . -j 2 --target real128_test;
cmake --build . -j 2 --target ecc_test;
cmake --build . -j 2 --target bip_lock;
echo "Build finished!"
```
The latest build log can be found at https://ci.appveyor.com/project/cwyyprog/bitshares-fc.
As far as I know, Ccache can save the compiling time, but it's still not enough to meet the 1-hour limit, due to ld.exe takes hours to link api.exe on my laptop.

### PR DESCRIPTION
**Pull request by pmconrad**
Fixes #1738 (hopefully)

The idea is to explicitly specialize certain templates (mostly exceptions, APIs and serialization) in `.cpp` files and declare them `extern` in `.hpp` files. This avoids them being instantiated (and thus compiled) separately whereever they are used.

The result is almost 40% faster build with a significant reduction of required RAM and diskspace. The effect on replay time seems to be lower than the error from measuring. 

Measurements:

Version | Compile time (full build with `-j 2`) |  Resulting total object file size | Replay time (32M blocks)
 ---- | ---- | ---- | ----
6a57bd9 (develop) | 4905.86user 142.54system 45:36.04elapsed 184%CPU (0avgtext+0avgdata 5910336maxresident)k 5883632inputs+36297912outputs (621major+79699588minor)pagefaults | 1930514440 | 4494s
275e5bfa2 (this branch) | 2995.77user 109.49system 27:34.96elapsed 187%CPU (0avgtext+0avgdata 3329756maxresident)k 400912inputs+30901272outputs (323major+60177298minor)pagefaults | 1129394680 | 4399s

**Note:** Requires changes in `fc` as well - https://github.com/bitshares/bitshares-fc/pull/133

### CODE REVIEW

**Comment by oxarbitrage** [comment]
nice! is the 40% build time reduction only in windows or all ?

**Comment by pmconrad** [comment]
Measured on my linux desktop. Don't have a windows machine to test with.

**Comment by abitmore** [comment]
@cwyyprog try this?

**Comment by pmconrad** [comment]
The fc bump includes the fix for #1688

**Comment by oxarbitrage** [comment]
I made a test and i was not able to get a 40% faster build in Ubuntu 18.04 but just a 15% .

develop branch:

```
real	77m46,590s
user	74m55,983s
sys	2m37,780s
```

pull request:

```
real	66m3,149s
user	63m3,007s
sys	2m32,247s
```

Also, did you made snapshots of the 2 at block 32M. I think it might worth it.

**Comment by pmconrad** [comment]
Snapshot comparison at 37.9M blocks produced empty diff.

File: libraries/wallet/include/graphene/wallet/wallet.hpp [hunk 1]
```diff
@@ -229,11 +229,12 @@ struct worker_vote_delta
    flat_set<worker_id_type> vote_abstain;
 };
 
-struct signed_block_with_info : public signed_block
+struct signed_block_with_info
 {
    signed_block_with_info( const signed_block& block );
    signed_block_with_info( const signed_block_with_info& block ) = default;
 
+   signed_block block;
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/wallet/include/graphene/wallet/wallet.hpp]: Good catch, thanks! I forgot to roll back that change after refactoring my changes.

**Pull request merged**